optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/rostergrid/view/ViewportView.java

```diff
@@ -39,21 +46,101 @@
     private ManagedInstance<LaneView<T>> laneViewInstances;
 
     @Inject
-    private ListView<Lane<T>> lanes;
+    private ListView<Viewport<T>, Lane<T>> lanes;
 
     @Inject
     private TimingUtils timingUtils;
 
+    @Inject
+    private PageUtils pageUtils;
+
+    @Inject
+    @Named("span")
+    private HTMLElement headerBackground;
+
+    private Viewport<T> viewport;
+
+    private boolean inMouseEvent = false;
+
     public void setViewport(final Viewport<T> viewport) {
 
         timingUtils.time("Viewport assemble", () -> {
+            this.viewport = viewport;
+            if (headerBackground.isConnected) {
+                headerBackground.remove();
+            }
             getElement().classList.add(viewport.decideBasedOnOrientation("vertical", "horizontal"));
 
-            viewport.setSizeInScreenPixels(this, viewport.getSizeInGridPixels(), 12L);
+            getElement().style.set("grid-template-columns", "auto " + "repeat(" + (viewport.getScale().getEndInGridPixels()) + ", 20px)");
 
-            viewport.drawTicksAt(() -> ticksLane);
+            pageUtils.expandElementToRemainingHeight(this);
 
-            lanes.init(getElement(), viewport.getLanes(), () -> laneViewInstances.get().withViewport(viewport));
+            headerBackground.classList.add("header-background");
+            getElement().appendChild(headerBackground);
+            viewport.setAbsPositionInScreenPixels(() -> headerBackground, 0L, 0L);
+            viewport.setSizeInScreenPixels(() -> headerBackground, viewport.getScale().getEndInGridPixels() + viewport.getHeaderColumns(), 0L);
+            viewport.setAbsGroupPosition(() -> headerBackground, 0);
+            viewport.setGroupSizeInScreenPixels(() -> headerBackground, viewport.getHeaderRows(), 0L);
+            viewport.drawTicksAt(this);
+            viewport.drawGridLinesAt(this);
+
+            lanes.init(getElement(), viewport, viewport.getLanes(), () -> laneViewInstances.get().withViewport(viewport));
         });
     }
+
+    @EventHandler("viewport")
+    private void onDragStart(@ForEvent("mousedown") MouseEvent e) {
+        if (!inMouseEvent && viewport != null && viewport.getMouseTarget() != null) {
+            try {
+                inMouseEvent = true;
+                e.target = viewport.getMouseTarget().getElement();
+                MouseEvent copy = new MouseEvent(e.type, (MouseEventInit) e);
+                viewport.getMouseTarget().getElement().dispatchEvent(copy);
```

Comment:
Why do you have to manually dispatch a copy of the event? Shouldn't the target receive the event already?

Reply:
Not always; consider the case you drag the mouse fast enough so that it is on the viewport but not the element, or if you are dragging the edge and you haven't move the mouse enough to increase the length.

Reply:
@mbarkley Correct; viewport.setMouseTarget(element) is called on element mouseDown event handler and viewport.setMouseTarget(null) is called on element mouseUp handler.